ThreadSafeObjectHeap lacks add and remove operations

kkinnunen-apple opened the pull request:
#### ecb76a449ffa26cb4da089afe4ee8d7ba90f13a9
<pre>
ThreadSafeObjectHeap lacks add and remove operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=257647">https://bugs.webkit.org/show_bug.cgi?id=257647</a>
rdar://110169358

Reviewed by Youenn Fablet.

Before, writing to ThreadSafeObjectHeap was intended to use the
retire(WriteReference, optional&lt;T&gt;, optional&lt;Timeout&gt;). The heap also
was intended to support copy-on-write or overwriting non-depending
update scenarios where an overwriting write could skip ahead the
earlier, non-processed reads and writes.

For example, receiving following sequence with reads omitted:
 write version=0, &quot;1&quot;    (sequence x)
 write version=1, &quot;abc&quot;  (sequence y)
 write version=2, &quot;abcd&quot; (sequence y)

In case of overwriting, non-depending writes, sequence y could proceed
first and insert version 1 and then subsequently version 2. Inserting
version 1 would insert a final read marker for version 0, which would be
noticed by eventual sequence x inserting first version 0. Upon
inserting version 0, the insert would be skipped as that version was
already overwritten.

However, since there was no distinction between &quot;write&quot; and &quot;initial
write&quot;, the final read marker would be needed for inserting any object,
including the initial insert. This in turn would mean that the heap
would be leaking the final read marker for the initial insertion.

Instead, change the API to &quot;add&quot;, &quot;get&quot;, &quot;read&quot;, &quot;take&quot;, &quot;remove&quot;. These
are more intuitive with the use-cases. Each write must naturally remove
the instance from the heap for the duration of the update, as would
de-facto happen in non-overwriting write also previously. Non-blocking
remove and add can be used for overwriting write until more performance
is needed.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::releaseSerializedImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::add):
(WebKit::RemoteVideoFrameObjectHeap::releaseVideoFrame):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.mm:
(WebKit::RemoteVideoFrameObjectHeap::get):
* Source/WebKit/Platform/IPC/ObjectIdentifierReferenceTracker.h:
(IPC::ObjectIdentifierReference::generateForAdd):
(IPC::ObjectIdentifierReferenceTracker::add const):
(IPC::ObjectIdentifierWriteReference::generateForAdd): Deleted.
* Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
(IPC::ThreadSafeObjectHeap::ReferenceState::ReferenceState):
(IPC::HeldType&gt;::add):
(IPC::HeldType&gt;::get):
(IPC::HeldType&gt;::read):
(IPC::HeldType&gt;::remove):
(IPC::HeldType&gt;::sizeForTesting const):
(IPC::HeldType&gt;::retire): Deleted.
(IPC::HeldType&gt;::retireRemove): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::properties):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/ThreadSafeObjectHeapTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/265063@main">https://commits.webkit.org/265063@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60a6ed634666b496dffcc9a17499621f46160953

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10004 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11164 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9324 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9513 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11741 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9720 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12232 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10534 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11322 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7828 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8651 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16070 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8923 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12136 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9288 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7553 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8490 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2326 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12713 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
```diff
@@ -53,35 +51,43 @@ class ThreadSafeObjectHeap {
+    // Waits until write is possible and removes the object or times out.
+    // To re-add after modification, the new reference can be obtained with
+    // WriteReference::retiredReference().
+    HeldType remove(WriteReference&&, Timeout);
```
It is more a take than a remove method.

kkinnunen-apple replied:
Good idea, renamed.

youennf commented on Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
```diff
@@ -53,35 +51,43 @@ class ThreadSafeObjectHeap {
-    void add(Identifier, HeldType&&);
+    // Establishes a new object in the heap with an initial reference.
+    // Returns true on success.
+    bool add(const Reference&, HeldType);
```
Why not HeldType&&?

kkinnunen-apple replied:
Passing via value is in most case more optimal, as this reduces the theoretical moves. The instance is created straight into the destination, not default-constructed and then moved to.

youennf commented on Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
```diff
@@ -91,7 +97,45 @@ class ThreadSafeObjectHeap {
+            m_objects.remove(addResult.iterator);
+            return true; // Reference was already removed, so nobody is waiting. No need to notify.
+        }
+        if (state.object)
```
We are not able to construct ThreadSafeObjectHeap that owns movable only type.
This check and other changes further enforce this, which is fine.
It would be nice to be able to have a ThreadSafeObjectHeap<Identifier, Ref<>> at some point.

kkinnunen-apple replied:
Added some HashTrait usage so that Ref works.
Note: The ..State::object is  optional<T>, if(state.object) works. 
Note: Ref is not move-only, it has copy operation. Ref is non-nullable.

Changed the tests to use Ref so it's more apparent.

youennf commented on Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
```diff
@@ -91,7 +97,45 @@ class ThreadSafeObjectHeap {
+}
+
+template<typename Identifier, typename HeldType>
+HeldType ThreadSafeObjectHeap<Identifier, HeldType>::read(ReadReference&& read, Timeout timeout)
```
Why is get taking a const& but read is taking a &&?
It seems both are making the same use of the read parameter.
Should we make get() const to better mark the difference?

kkinnunen-apple replied:
get() does not consume the read reference.
read() consumes the read reference, it cannot be used after.
Marked as const

kkinnunen-apple committed:
ThreadSafeObjectHeap lacks add and remove operations
https://bugs.webkit.org/show_bug.cgi?id=257647
rdar://110169358

Reviewed by Youenn Fablet.

Before, writing to ThreadSafeObjectHeap was intended to use the
retire(WriteReference, optional<T>, optional<Timeout>). The heap also
was intended to support copy-on-write or overwriting non-depending
update scenarios where an overwriting write could skip ahead the
earlier, non-processed reads and writes.

For example, receiving following sequence with reads omitted:
 write version=0, "1"    (sequence x)
 write version=1, "abc"  (sequence y)
 write version=2, "abcd" (sequence y)

In case of overwriting, non-depending writes, sequence y could proceed
first and insert version 1 and then subsequently version 2. Inserting
version 1 would insert a final read marker for version 0, which would be
noticed by eventual sequence x inserting first version 0. Upon
inserting version 0, the insert would be skipped as that version was
already overwritten.

However, since there was no distinction between "write" and "initial
write", the final read marker would be needed for inserting any object,
including the initial insert. This in turn would mean that the heap
would be leaking the final read marker for the initial insertion.

Instead, change the API to "add", "get", "read", "take", "remove". These
are more intuitive with the use-cases. Each write must naturally remove
the instance from the heap for the duration of the update, as would
de-facto happen in non-overwriting write also previously. Non-blocking
remove and add can be used for overwriting write until more performance
is needed.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::releaseSerializedImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::moveToSerializedBuffer):
(WebKit::RemoteRenderingBackend::moveToImageBuffer):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::add):
(WebKit::RemoteVideoFrameObjectHeap::releaseVideoFrame):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.mm:
(WebKit::RemoteVideoFrameObjectHeap::get):
* Source/WebKit/Platform/IPC/ObjectIdentifierReferenceTracker.h:
(IPC::ObjectIdentifierReference::generateForAdd):
(IPC::ObjectIdentifierReferenceTracker::add const):
(IPC::ObjectIdentifierWriteReference::generateForAdd): Deleted.
* Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h:
(IPC::ThreadSafeObjectHeap::ReferenceState::ReferenceState):
(IPC::HeldType>::add):
(IPC::HeldType>::get):
(IPC::HeldType>::read):
(IPC::HeldType>::remove):
(IPC::HeldType>::sizeForTesting const):
(IPC::HeldType>::retire): Deleted.
(IPC::HeldType>::retireRemove): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::properties):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/ThreadSafeObjectHeapTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/265063@main
